website/blog/2024-10-28-kustomize-vs-helm/index.mdx

```diff
@@ -0,0 +1,179 @@
+
+### Learning Curve and Maintainability
+
+Helm has a steeper learning curve due to its templating system and the need to understand the Helm-specific syntax. Kustomize, being closer to native Kubernetes, might feel more intuitive for users already familiar with Kubernetes resource definitions. However, this simplicity can also lead to challenges when dealing with more complex configurations.
+
+## Using Helm and Kustomize Together
+
+Helm and Kustomize can each be used as standalone tools, but in practice, they are often combined to manage Kubernetes deployments more effectively. Consider using them together in the following scenarios:
+
+-   **Environment-specific customizations**: Use Kustomize to apply changes such as resource limits, quotas, annotations, or namespaces across multiple environments (e.g., development, staging, production) without altering the original Helm chart. This allows you to dynamically configure any environment-specific adjustments while maintaining the core deployment logic in Helm, ensuring that the base setup remains consistent across environments.    
+-   **When using third party Helm charts**: Modify manifests from third-party Helm charts without altering the original chart using Kustomize (e.g., adding labels or changing configurations).    
+-   **Managing Secrets and ConfigMaps**: Keep sensitive resources separate from Helm charts by creating them with Kustomize after the chart is deployed. Kustomize can also be used to inject sensitive resources that have already been created.   
+-   **GitOps workflows:** Use Helm to generate the base manifests, keep Helm rendered manifests and environment customizations separate while adhering to a clean, declarative GitOps model. Using Helm and Kustomize together supports the principle of separation of concerns, which is a key aspect of the GitOps methodology.   
+
+There are two main, mutually incompatible approaches to combining Helm and Kustomize.
+
+1.  **Using** `helm template` **and apply with Kustomize**: In this method, you use `helm template` to generate the Kubernetes manifests from a Helm chart, and then use Kustomize to apply patches or overlays. Since Kustomize is in charge of deploying the resources to Kubernetes, Helm won’t manage the release, so there’s no Helm-based release tracking.
+    
+2.  **Using** `helm install` **with Kustomize pre-modifications**: Through this method, Helm is responsible for the application release so it manages and keeps track of the application's lifecycle, while Kustomize modifies the manifests before applying them to the cluster. This allows you to still benefit from Helm’s release management.
```

Comment:
I think it would be fair to also mention the "HelmChartInflationGenerator" built into kustomize, although it seems to have some issues and it's future is a bit uncertain: 
https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/

Reply:
Nice one! I wasn't aware of this